### DISCUSSION TITLE
On definition of generalisation error in last part of autodiff exercise (Last bit of Task 5)

### DISCUSSION THREAD

**Discussion by tony-teng-yt377** [Part-II-Computational-Physics/.github#74] [General]
Hi, I am currently at generalisation error in last part. The test sets are five blue and five red circles. I wonder does the generalisation error basically evaluated as follows:

1. Plot the boundary generated by each method.
2. Count the error (if red points of test set are located in blue region of predicted boundary or blue points in red region, then count as true?)
3. Compare the errors of three methods?

**Comment by Jamiewhbee** [comment]
From my understanding, the generalisation error is the accuracy of a model when applied to an unseen set of data (the test set), so this method does correctly evaluate the generalisation error. You can use a previously defined function to calculate the accuracy.

**Reply by tony-teng-yt377** [reply]
Thanks! Will implement it soon!

**Comment by ceciliegl** [comment]
Hi @tony-teng-yt377 and @Jamiewhbee ! The generalisation error is defined as the difference between the loss function evaluated on the test set and the training set (https://en.wikipedia.org/wiki/Generalization_error).

**Reply by JordanEmil** [reply]
Hello, are you sure this is correct? The wikipedia page you have linked seems to contradict your statement about the generalisation error. 
![WhatsApp Image 2023-03-14 at 1 15 53 PM](https://user-images.githubusercontent.com/91506272/225014391-7b212b67-a0d3-48f1-875d-14857081dda6.jpeg)


You can't compute the expectation exactly because you don't know the data distribution, but you use the test samples as an approximation to it

**Reply by ceciliegl** [reply]
I copied that definition and the link from the lecture notes. Sorry if it was confusing. The generalisation error is in practice measured by the difference between the loss of the training set and the loss of the test set.

**Reply by AustenLamacraft** [reply]
@JordanEmil the expression you've circled can't be evaluated (as the article explains) so you use the empirically averaged loss functions instead.

**Reply by tony-teng-yt377** [reply]
So are we meant to only compare loss functions at the end or compare it at every step? (I guess in latter case we can see whether the model gets overtrained?)

**Reply by tony-teng-yt377** [reply]
![df006f2b-cb76-4762-8083-745aa0c93cbc](https://user-images.githubusercontent.com/125870751/225110084-54d4d8a9-f8a6-4c64-94ec-a2c089992ef5.png)
So I just look at the difference between the loss function for test set (using trained model) and the loss function of training as a function of steps. I guess if it is overtrained, there should be a min followed by a growth, but it seems that except for Adam potentially overtrained, both gd and Lion are undertrained. Is that what we mean to do?
